src/MongrationDotNet/ClientSideDocumentMigration.cs

```diff
@@ -29,7 +28,10 @@ public override async Task ExecuteAsync(IMongoDatabase database, ILogger logger)
             IEnumerable<BsonDocument> documents;
             do
             {
-                documents = (await collection.Find(SearchFilters).ToListAsync()).Skip(updated).Take(BatchSize);
+                documents = PageThroughAllFilteredDocuments 
+                    ? (await collection.Find(SearchFilters).ToListAsync()).Skip(updated).Take(BatchSize).ToArray() 
+                    : (await collection.Find(SearchFilters).ToListAsync()).Take(BatchSize).ToArray();
+                
```

Comment:
If you compare with the version i created internally you can see I do the Take() in the database query rather than as here where you are loading the entire search into memory and then throwing most of it away:

```
                documents = await collection.Find(SearchFilters).Limit(BatchSize).ToListAsync();
```

For the version where PageThroughAllFilteredDocuments = false the paging does not make sense - you have already copied the entire database collection into RAM, what's the point in paging?!
```

Reply:
Also upsert should never be allowed - it MUST be an update, and insert would be very bad.
` ReplaceOptions {IsUpsert = true});`

Reply:
Fixed